tests/test_pulp_handler.py

```diff
@@ -0,0 +1,96 @@
+import logging
+import os
+
+import mock
+from pubtools.pluggy import pm  # pylint: disable=wrong-import-order
+from six.moves.urllib.parse import urljoin
+
+from pubtools.exodus.hooks import pulp
+
+MOCK_ENV = {
+    "EXODUS_ENABLED": "True",
+    "EXODUS_GW_URL": "https://exodus-gw.example.com",
+    "EXODUS_GW_ENV": "test",
+    "EXODUS_GW_CERT": "path/to/ex.crt",
+    "EXODUS_GW_KEY": "path/to/ex.key",
+}
+
+
+@mock.patch.dict(os.environ, MOCK_ENV)
+def test_exodus_pulp_hanlder(requests_mock, caplog):
+    caplog.set_level(logging.DEBUG, "pubtools-exodus")
+
+    # Construct handler
+    auth_resp = {
+        "client": {
+            "roles": [],
+            "authenticated": False,
+            "serviceAccountId": "test-client",
+        },
+        "user": {
+            "roles": ["tester"],
+            "authenticated": True,
+            "internalUsername": "test-user",
+        },
+    }
+    auth_url = urljoin(MOCK_ENV["EXODUS_GW_URL"], "/whoami")
+    requests_mock.get(auth_url, json=auth_resp, status_code=200)
+
+    # Simulate task start
+    pulp.task_start()
+
+    assert (
+        "You are authenticated as user test-user with roles: ['tester']"
+        in caplog.text
+    )
+
+    # Simulate repo publish; call prep-publish hook
+    publish_resp = {
+        "id": "497f6eca-6276-4993-bfeb-53cbbbba6f08",
+        "env": "test",
+        "links": {
+            "self": "/test/publish/497f6eca-6276-4993-bfeb-53cbbbba6f08",
+            "commit": "/test/publish/497f6eca-6276-4993-bfeb-53cbbbba6f08/commit",
+        },
+        "items": [],
+    }
+    publish_url = os.path.join(
+        MOCK_ENV["EXODUS_GW_URL"], MOCK_ENV["EXODUS_GW_ENV"], "publish"
+    )
+    requests_mock.post(publish_url, json=publish_resp, status_code=200)
+    to_publish = pm.hook.pulp_repository_pre_publish(
+        to_publish=[(None, {"rsync_extra_args": ["--existing-arg"]})]
+    )
+    assert "Created publish: %s" % publish_resp["id"] in caplog.text
+
+    # The pre-publish hook should have appended exodus-publish arg to
+    # existing rsync_extra_args.
+    # TODO: why is hook returning additional list dimention
```

Comment:
Because there can be more than one hook implementer.  You get a list of results and not just one result, so you can get the result of all hookimpls if there are more than one.  https://pluggy.readthedocs.io/en/stable/#collecting-results